### DISCUSSION TITLE
Present your Excalidraw sketch-note slide in a "Prezi" like style

### DISCUSSION THREAD

**Discussion by zsviczian** [zsviczian/obsidian-excalidraw-plugin#140] [Show and tell]
![Step-by-step guide title](https://user-images.githubusercontent.com/14358394/132868931-9aa47b8c-2b65-44da-b38f-ea6e4b6aeb61.png)

See this 📽[video](https://youtu.be/fWuFeMMZid4) for a walkthrough of this example.

# Plugins used in the example (beyond Excalidraw of course)
### CodeView
I use my [CodeView plugin](https://github.com/zsviczian/obsidian-codeeditor/releases/tag/0.0.3) for editing javascript files within Obsidian. This plugin is not yet available in community plugins, but you can install it following the link. Also, you can use an external editor to edit javascript files (e.g. notepad++).

### QuickAdd
[QuickAdd](https://github.com/chhoumann/quickadd) is available in community plugins.

# The scripts
QuickAdd reads `*.js` scripts from the "scripts" folder of your Vault. Create the "scripts" folder and then two .js files in it. Copy the contents of these files into the two files respectively. You may need to use CTRL+SHIFT+V to paste to preserve formatting. Note, that unless you have a plugin like CodeView installed, you will not be able to create `.js` files inside Obsidian (but you will need to open the folder in your system's file explorer). Files you create in Obsidian will always get a `.md` extension by default. If you have CodeView installed, you can right-click the scripts folder and select "Add js file".

You can also download the [5 js scripts](https://github.com/zsviczian/obsidian-excalidraw-plugin/files/7143817/prezi-scripts.zip) from here, unzip them and drop them into the scripts folder. Note, that you need to enable Obsidian Settings / Files & Links / Detect all file extensions to see the files you have dropped in the folder.

### Prezi Capture groupId.js
```javascript
//capture the groupId of the selected group
module.exports = (params) => {
	ea = ExcalidrawAutomate;
	ea.setView("active");
	elements = ea.getViewSelectedElements();
	
	//count number of element in each group
	groupIds = {};
	elements.forEach( (el) => el.groupIds.forEach( (gId)=>{
		if(!(gId in groupIds)) groupIds[gId] = 0;
		groupIds[gId]++;
	}));
	
	//select group with highest number of elements
	groupId = ["",0];
	for(key of Object.keys(groupIds)) {
		if(groupIds[key] > groupId[1]) {
			groupId = [key,groupIds[key]];
		}
	}
	params.variables["groupId"] = groupId[0];
	new Notice("Capture OK",800);
}
```

### Prezi Get navigation-file path.js
```javascript
//calculate the filepath for the file containing the slide (groupId) sequence for the presentation
module.exports = (params) => {
	const path = app.workspace.activeLeaf.view.file.path;
	const lastDot = path.lastIndexOf(".");
	const length = (lastDot === -1) ? path.length : lastDot;
	params.variables["navigationFilePath"] = `${path.substr(0,length)} - Navigation`;
	params.variables["direction"] = "forward";
}
```

### Prezi Start presentation.js
```javascript
module.exports = (params) => {
	const ea = ExcalidrawAutomate;
	ea.setView("active");
	if(!document.fullscreenElement) {
		ea.viewToggleFullScreen(true);
	}
	window.localStorage.setItem("previousSlide", -1);
}
```

### Prezi Previous slide.js
```javascript
module.exports = (params) => {
	params.variables["direction"] = "back";
}
```

### Prezi Display slide.js
```javascript
//Displays the slide specified in nextSlide and increments nextSlide (or if end, sets to beginning)
module.exports = async (params) => {
	const ea = ExcalidrawAutomate;
	ea.setView("active");
	
	const navigationFilePath = params.variables["navigationFilePath"];
	if(!navigationFilePath) return;
	const f = app.metadataCache.getFirstLinkpathDest(navigationFilePath,"");
	slides = (await app.vault.cachedRead(f)).split("\n"); //create an array of groupIds 
	
	let slideNumber = parseInt(window.localStorage.getItem("previousSlide"));
	if(params.variables["direction"] === "forward") {
		slideNumber = (slideNumber<slides.length-1) ? slideNumber+1 : 0;
	} else {
		slideNumber =  (slideNumber>0) ? slideNumber-1 : slides.length-1;
	}
	window.localStorage.setItem("previousSlide",slideNumber);
	
	
	slide  = slides[slideNumber];

	//Zoom to slide
	let elements = ea.getExcalidrawAPI().getSceneElements();
	if(slide!="") {
  		elements = elements.filter((el)=>el.groupIds.includes(slide));
	}
	ea.getExcalidrawAPI().zoomToFit(elements,10);
}
```

# QuickAdd config
[QuickAdd-data.json.zip](https://github.com/zsviczian/obsidian-excalidraw-plugin/files/7143864/QuickAdd-data.json.zip)
If this is your only (first ever) QuickAdd automation in your Vault, you can navigate to `.obsidian\Plugins\quickadd` and replace the data.json file with the one in the zip file. 
- Install QuickAdd from community plugins
- Enable QuickAdd
- Exit Obsidian
- Replace the data.json file with the one provided
- Start Obsidian
- Set up hotkeys if preferred

**Comment by Madd0g** [comment]
I was wondering about how to achieve something like this, very cool.

Question, since the plugin can store stuff inside the excalidraw file itself, wasn't it possible to store the list of ids inside the file instead of requiring an additional external file?

Also wondering if it's possible to animate the transition to the next area instead of the immediate jump. I couldn't find how to use `zoomToFit` from the excalidraw github.
